File: src/libkuai/kuai/mols/PBC.cpp
#include <kuai/mols/PBC.h>


namespace kuai {

	PBC::PBC() {
		ax = by = cz = LARGE_NUMBER;
		bx = cx = cy = 0;
	}

	PBC::PBC(const RealNumber& a, const RealNumber& b, const RealNumber& c) 
	{
		ax = a;
		by = b;
		cz = c;
		bx = cx = cy = 0;
	}

	PBC::PBC(const RealNumber& a, const RealNumber& b, const RealNumber& c, const RealNumber& alpha, const RealNumber& beta, const RealNumber& gamma)
	{
		RealNumber cosA = cos(alpha), cosB = cos(beta), cosG = cos(gamma);
		RealNumber sinG = sin(gamma);
		ax = a;

		bx = b * cosG;
		by = b * sinG;

		cx = c * cosB;
		cy = c * (cosA - cosB*cosG) / sinG;
		cz = sqrt(c*c - cx*cx - cy*cy);
	}

	const RealNumber PBC::a() const {
		return ax;
	}
	const RealNumber PBC::b() const {
		return sqrt(bx*bx + by*by);
	}
	const RealNumber PBC::c() const {
		return sqrt(cx*cx + cy*cy + cz*cz);
	}

	const XYZ PBC::va() const {
		return XYZ(ax, 0, 0);
	}

	const XYZ PBC::vb() const {
		return XYZ(bx, by, 0);
	}

	const XYZ PBC::vc() const {
		return XYZ(cx, cy, cz);
	}

	const RealNumber PBC::alpha() const {
		return angle(vb(), vc());
	};
	const RealNumber PBC::beta() const {
		return acos(cx/c());
	}
	const RealNumber PBC::gamma() const {
		return atan2(by, bx);
	}

	void PBC::update(XYZ& v) const {
		while (v.z < -cz/2) {
			v.x += cx;
			v.y += cy;
			v.z += cz;
		}
		while (v.z > cz/2) {
			v.x -= cx;
			v.y -= cy;
			v.z -= cz;
		}
		while (v.y < -by/2) {
			v.x += bx;
			v.y += by;
		}
		while (v.y > by/2) {
			v.x -= bx;
			v.y -= by;
		}
		while (v.x < -ax/2) {
			v.x += ax;
		}
		while (v.x > ax/2) {
			v.x -= ax;
		}
	}

	const XYZ PBC::image(const XYZ& v) const {
		XYZ result(v);
		update(result);
		return result;
	}

	RealNumber PBC::volumn() const {
		return ax * by * cz;
	};

	void PBC::scale_x(RealNumber factor) {
		ax *= factor; bx*= factor; cx *= factor;
	};
	void PBC::scale_y(RealNumber factor) {
		by *= factor; cy *= factor;
	};
	void PBC::scale_z(RealNumber factor) {
		cz *= factor;
	};

	void PBC::scale(RealNumber x, RealNumber y, RealNumber z) {
		scale_x(x);
		scale_y(y);
		scale_z(z);
	}
	
	void PBC::scale(RealNumber v) {
		scale(v, v, v);
	}

	PBC PBC::parse(const String& v0) {
		StringArray tokens = split(v0);
		if (tokens.size() >= 6) {
			try {
				Index offset = tokens.size() - 6;
				RealNumber x = lexical_cast<RealNumber>(tokens[offset+0]);
				RealNumber y = lexical_cast<RealNumber>(tokens[offset+1]);
				RealNumber z = lexical_cast<RealNumber>(tokens[offset+2]);

				RealNumber alpha = lexical_cast<RealNumber>(tokens[offset+3]);
				RealNumber beta  = lexical_cast<RealNumber>(tokens[offset+4]);
				RealNumber gamma = lexical_cast<RealNumber>(tokens[offset+5]);

				alpha = deg2rad(alpha);
				beta  = deg2rad(beta);
				gamma = deg2rad(gamma);

				return PBC(x, y, z, alpha, beta, gamma);
			}
			catch (BadLexicalCast&) 
			{ }
		}
		throw BadLexicalCast(v0, typeid(PBC));
	}

}

File: gromacs/basicTools/pbc.py
from xyz import *
from top import angle
from math import cos, acos, sin, degrees, atan2, radians, pi


class PBC:
    """ class for the  Protein Data Bank (PDB) format
    """
    def __init__(self, a, b, c, alpha=pi/2, beta=pi/2, gamma=pi/2):
        """@param a: box a
        @param b: box b
        @param c: box c
        @param alpha: alpha angle default is pi/2
        @param beta: alpha angle default is pi/2
        @param gamma: alpha angle default is pi/2
        """
        cosA, cosB, cosG = cos(alpha), cos(beta), cos(gamma)
        """@var cosA: value of cos(alpha)
        @var cosB: value of cos(beta)
        @var cosG: value of cos(gamma)
        """
        sinG = sin(gamma)
        """@var sinG: value of sin(gamma)"""
        self.ax = a
        """@var ax: xx tensor"""
        self.bx,  self.by = b*cosG, b*sinG
        """@var bx: yx tensor"""
        """@var bx: yy tensor"""
        self.cx, self.cy, = c*cosB, c*(cosA-cosB*cosG)/sinG
        """@var cx: zx tensor"""
        """@var cy: zy tensor"""
        self.cz = sqrt(c*c-self.cx*self.cx - self.cy*self.cy)
    
    @property
    def a(self):
        return self.ax
    
    @property
    def b(self):
        return sqrt(self.bx*self.bx + self.by*self.by)
    
    @property
    def c(self):
        return sqrt(self.cx*self.cx + self.cy*self.cy + self.cz*self.cz)
    
    @property
    def va(self):
        return XYZ(self.ax, 0.0, 0.0)
    
    @property
    def vb(self):
        return XYZ(self.bx, self.by, 0.0)
    
    @property
    def vc(self):
        return XYZ(self.cx, self.cy, self.cz)
    
    @property
    def alpha(self):
        return angle(self.vb, self.vc)
        
    @property
    def beta(self):
        return acos(self.cx/self.c)
    
    @property
    def gamma(self):
        return atan2(self.by, self.bx)
    
    @property
    def volumn(self):
        return self.ax*self.by*self.cz
    
    def scale(self, v):
        try:
            self.x *= v.x
            self.y *= v.y
            self.z *= v.z
        except AttributeError:
            self.x *= v
            self.y *= v
            self.z *= v
            
    def norm(self, xyz):
        x = xyz.x
        y = xyz.y
        z = xyz.z
        while z < -self.cz/2:
            z += self.cz
            y += self.cy
            x += self.cx
        while z > self.cz/2:
            z -= self.cz
            y -= self.cy
            x -= self.cx
        while y < -self.by/2:
            y += self.by
            x += self.bx
        while y > self.by/2:
            y -= self.by
            x -= self.bx
        while x < -self.ax/2:
            x += self.ax
        while x > self.ax/2:
            x -= self.ax

        return XYZ(x, y, z)

    @staticmethod
    def parse(line):
        tokens = line.split()
        if len(tokens) >= 6:
            try:
                tokens = [float(i) for i in tokens[-6:]]
                tokens[3] = radians(tokens[3])
                tokens[4] = radians(tokens[4])
                tokens[5] = radians(tokens[5])
                return PBC(tokens[0], tokens[1], tokens[2], tokens[3], tokens[4], tokens[5])
            except:
                pass
        return None
            
        
    def __str__(self):
        return "PBC: %12.4f %12.4f %12.4f %5.2f %5.2f %5.2f" % \
                (self.a, self.b, self.c, 
                 degrees(self.alpha), degrees(self.beta), degrees(self.gamma))
